Declarative refactoring

thomcorley opened the pull request:
All modules are now using a more-declarative approach in which all the
information to create the whole section (eg: ingredients list, method
steps etc) is being returned as an object by the module, and is then
being rendered in recipe.js, with the parent element being passed in.

There is one exception; metadata.js, which still renders more than once
within the module. This is due to some complexity with sibling elements
and will be addressed in a future PR.

thomcorley committed:
Tidy up recipe.js

Extracts the logic for displaying the various components of a recipe
into separate files

thomcorley committed:
Refactor the recipe.js code to be more declarative

* reworked image.js
* refactored the metadata script to be more decarative

thomcorley committed:
Add eslint and save Shane's scratchpad

ShaneDrury commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,54 @@
+    ]
+  }
+
+  let publishedLabel = (published, selector) => {
```
this `let` can be a `const`

ShaneDrury commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,54 @@
+    });
+  }
+
+  let populateTags = (tagNames, parentElement) => {
```
If we controlled the parent element it could look something like:

```js
const span = ({ className, children, innerHTML }) => ({
  element: "span",
  className,
  children,
  innerHTML,
})

const tags = tagNames.map(tagName => span({ innerHTML: `${tagName},&nbsp` }));
const parent = span({ className: "tags", children: tags });

const render = (target, { element, className, children, innerHTML }) => {
  const el = document.createElement(element);
  el.innerHTML = innerHTML; // TODO: only if innerHTML is not undefined
  el.className = className;
  el.innerHTML = children.map(child => { render(el, child); }); // if children defined - possibly only if innerHTML is not defined
  target.appendChild(el);
  return el;
}

render(document.body, parent);
```

Possibly if we use JSX that could become:

```jsx
const tags = tagNames.map(tagName => (<span innerHTML={`${tagName},&nbsp`} />));
const parent = <span className="tags" children={tags} />
```

thomcorley committed:
Sketched out a plan for declarative refactoring

* Added `render` method and used it to generate the tags

thomcorley committed:
WIP: address TODOs

thomcorley committed:
Reworked the metadata file with the new declarative approach

thomcorley committed:
Use declarative approach for image file

thomcorley committed:
Adds declarative approach for introduction

thomcorley committed:
Remove the now ignored directory node_modules

ShaneDrury commented on js/recipe/introduction.js:
```diff
@@ -0,0 +1,16 @@
+  const firstParagraphDiv = els.div({className: "first-paragraph", children: [firstParagraphElement]});
+  const introductionParentElement = render(recipeParentElement, els.div({className: "introduction-paragraphs", children: [firstParagraphDiv]}));
+
+  latterDescriptionParagraphs.forEach(paragraph => {
```
I wonder if this may rely on the fact that so far we don't clear the targetElement when rendering to it. One requirement we may find very useful about `render` later is that it is idempotent i.e. the html doesn't change if you call it twice with the same things. That'll be useful later if we decide to re-render automatically when the inputs change (say as the result of an API response).
Is there another way to structure this that uses `map` instead, for example? Or something that calls `render` once?

thomcorley replied:
I am still working on getting `render` to not repeat the html when it is called twice... but I've managed to update this introduction.js code so that it only calls render once, with all the children of the target node passed in in one go, which is pretty neat!

ShaneDrury commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,31 @@
+
+export function metadata(recipe) {
+  // Recipe Data
+  const tags = recipe.tags;
```
Might be a nice candidate for object destructuring e.g.:
```js
export function metadata({ tags, datePublished, recipeYield, ...}) {
```
although it can make the parameter list a bit noisier.

thomcorley replied:
Yes! (just read up on Object destructuring), that is super neat!

ShaneDrury commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,31 @@
+  const recipeParentElement = document.querySelector("article.post.single");
+
+  // TITLE
+  render(recipeParentElement, htmlElement("h1")({className: "post-title", innerHTML: title}));
+
+  // METADATA
+  const metaParentElement = render(recipeParentElement, els.div({className: "post-meta clear"}));
```
I think this is also relying on the fact that `render`-ing to the same target element twice puts the elements next to each other. I'd suggest we change `render` to clear the target element first, then think about how we may declare that `h1` and `div` sit side by side - calling `render` once.
Ideally we could only call `render` a handful of times in all of this JS (similar to Banking React code), but we can gradually move towards that.

thomcorley replied:
Yes, I can use the same approach that I used in introduction.js to render it once, with all the children passed in in one go

thomcorley replied:
I've managed to get most of the metadata elements rendered at once by creating an array of `metaChildren` elements and passing them in when I render the parent div. The title element is not included in this however because it's not a child node of the parent metadata element.

I'm wondering how to do this and I think we could modify render to take a `siblings` argument, and then use appendChild to add those to the target at the end of the render method?

ShaneDrury commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,31 @@
+
+  // Date published
+  const datePublishedElement = render(metaParentElement, els.time({innerHTML: publishedAt}));
+  datePublishedElement.setAttribute("datetime", "2020-12-16T09:09:00+00:00");
```
Can this be injected into the `els.time({innerHTML: publishedAt})` declaration? We can have specific elements take additional parameters without adding all of those parameters to `htmlElement`explicitly in one of two ways - let `htmlElement` take arbitrary parameters that we set dynamically, or extend `els.time` to also take `datetime`. I'd vote for the second to be more explicit.

thomcorley replied:
Yes, agreed. I've added an argument to els.time 👌

thomcorley replied:
Hmm, I cannot get this to work correctly, I will ask you about it later today.

ShaneDrury commented on js/recipe/image.js:
```diff
@@ -0,0 +1,16 @@
+import { htmlElement } from "../htmlRenderer.js"
+import { render } from "../htmlRenderer.js"
+
+export function image(recipe) {
```
This is a nice candidate for a self-contained "component". We can think about what that would take as arguments (in this case probably just `image` and `name`). I'd then suggest moving the `document.querySelector` and `render` calls out of this function and into the top-level `recipe.js` which calls this. That would be in the spirit of separating out the declaration and interpretation, and makes this a bit more self-contained.
In that case this function would end:
```
return htmlElement("img")({ attributes: imageAttributes });
```
So we return an object that declares what an `image` is, rather than calling it, expecting it to `render` the image for us.

ShaneDrury replied:
Possibly a component for image (or recipeImage) would look like:

```js
const image = ({ image, name }) => {
  return htmlElement("div")({className: "recipe-image"});
}
```

thomcorley replied:
Nice. I've moved the rendering of this section into recipe.js 👌

ShaneDrury reviewed (commented):
Nice work!

ShaneDrury commented on js/htmlRenderer.js:
```diff
@@ -0,0 +1,51 @@
+  attributes
+});
+
+const els = {
```
```js
export const span = htmlElement("span");

//then later
export els = {
  span,
  ...
};
```
Import styles:
```
import * from "render.js" // unqualified import
import * as els from "render.js" // qualified import els.render
import { span, els } from "render.js" // object destructuring import
```

ShaneDrury commented on js/recipe/image.js:
```diff
@@ -0,0 +1,16 @@
+
+  const recipeParentElement = document.querySelector("article.post.single");
+  const imageDiv = render(recipeParentElement, htmlElement("div")({className: "recipe-image"}));
+  render(imageDiv, htmlElement("img")({ attributes: imageAttributes }));
```
I believe that this can be passed in as:

```js
htmlElement("div")({className: "recipe-image", children: htmlElement("img")({ attributes: imageAttributes }) });
```
and `render` once

ShaneDrury replied:
This is an example of a "component-tree", where each component may or may not contain other components.

thomcorley replied:
Yep. I did the same thing when I was refactoring metadata.js, only calling `render` once and then passing all the child elements in at the same time. There is no need for this now however, since I am rendering the image component from recipe.js

ShaneDrury commented on js/recipe/recipe.js:
```diff
@@ -0,0 +1,30 @@
+
+function displayRecipe(recipe) {
+  metadata(recipe);
+  image(recipe);
```
If `image` were a component, this could be `render(document.querySelector("article.post.single"), image(recipe))` or similar.

thomcorley replied:
Yes! This is what it looks like now actually:
```
render(recipeParentElement, image(recipe));
```

As I have extracted `recipeParentElement` into a constant file.

thomcorley committed:
Adds node_modules folder to the gitignore

thomcorley committed:
Use * for a cleaner import syntax

thomcorley committed:
Moves implementation of image into recipe.js

thomcorley committed:
Various declarative refactorings

thomcorley committed:
Refactoring for the metadata class

thomcorley committed:
Improve the time method declaration to take an additional arg

thomcorley committed:
Refactor image.js

thomcorley commented on js/htmlRenderer.js:
```diff
@@ -0,0 +1,51 @@
+
+
+  target.appendChild(el);
+
```
If there were siblings passed in to render, then I could append them here:
```
siblings.forEach(sibling => {target.appendChild(sibling})

thomcorley commented on js/recipe/metadata.js:
```diff
@@ -0,0 +1,27 @@
+
+  const metaChildren = [
+    els.time({innerHTML: date}),
+    els.span({
+      className: "tags",
+      innerHTML: "/ Tags: &nbsp",
+      children: [
+        els.span({children: tagElements})
+      ]
+    }),
```
I cannot get this tags section to display properly, I would like to pair with you on it and figure out what I'm missing!

thomcorley replied:
Solution: I was using innerHTML _and_ children which is not allowed in our implementation.

ShaneDrury commented on js/htmlRenderer.js:
```diff
@@ -0,0 +1,55 @@
+const clearElements = (elements) => {
+  while (elements.firstChild) {
+    elements.removeChild(element.lastChild);
```
Looks pretty similar to how React does it 👍  
https://github.com/facebook/react/blob/0203b6567c6fd6274866c853ef938241d24551ec/packages/react-dom/src/client/ReactDOMLegacy.js#L123

thomcorley replied:
Nice!

ShaneDrury commented on js/recipe/introduction.js:
```diff
@@ -0,0 +1,17 @@
+import * as els from "../htmlRenderer.js"
+
+export function introduction(recipe) {
+  const firstParagraphElement = els.p({innerHTML: recipe.description[0]});
+  const firstParagraphDiv = els.div({className: "first-paragraph", children: [firstParagraphElement]});
+  const latterDescriptionParagraphs = recipe.description.slice(1);
+
+  const latterParagraphElements = latterDescriptionParagraphs.map(paragraph => {
+    return els.p({innerHTML: paragraph});
+  });
+
+  const children = [firstParagraphDiv].concat(latterParagraphElements);
+
+  render(recipeParentElement, els.div({className: "introduction-paragraphs", children: children}));
```
Might be interesting to see what this looks like without the intermediate variables:

```js
els.div({ className: "introduction-paragraphs", children: [
    els.div({ className: "first-paragragh", children: [
      els.p({ innerHTML: recipe.description[0] })
      ]}),
    ...latterDescriptionParagraphs.map(paragraph => els.p({ innerHTML: paragraph })
    )
  ]});
```

(taking some liberties with indentation) and compare it to the JSX version:

```jsx
<div className="introduction-paragraphs">
  <div className="first-paragraph">
    <p>{recipe.description[0]}</p>
  </div>
  {latterDescriptionParagraphs.map(paragraph => (<p>{paragraph}</p>))
</div>
```

This also suggests to me that always needing `children` to be a list is annoying and could be changed (we could check inside `render` if `children` is a list and wrap it if not).

ShaneDrury replied:
And it might be useful to think about how the top thing would be implemented in Ruby where we have blocks:

```ruby
div(className: "introduction-paragraphs") do
  div(className: "first-paragraph") do
    p { recipe.description[0] }
  end
  ...
end
```

We've actually implemented a DSL in JS, which is sometimes more natural in Ruby because it has blocks (which allow us to express parent-child relationships like HTML elements).

thomcorley replied:
That's really cool that you can see the similarities with JSX! I suppose it's not too surprising since you have been nudging us along gradually to a React-like framework :)

thomcorley replied:
When you say it's annoying that children has to be a list and we could put a check in `render`, do you mean that we could pass an array of arrays to `render` and it would wrap them into a single array?

thomcorley replied:
Also - is "list" the correct term here rather than "array" ?

ShaneDrury replied:
Ah yeah I'm just being loose with array/list, it's an [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array) in JS.
re: children in a list, I was thinking that it's quite annoying to always wrap them in a list, even if there's one child e.g.:

```js
els.div({ className: "foo", children: [els.div({ className: "bar" })] })
```

We could allow `render` to accept `children` as either being an element or an array of elements. That would mean `render` has to check the type of `children` and handle each case slightly differently.

Then again, it's no big deal to wrap things in a list really :)

thomcorley replied:
I'm going to make a note to change this behaviour of the children in htmlRenderer.js in a future PR, and get this one merged.

ShaneDrury commented on js/recipe/introduction.js:
```diff
@@ -0,0 +1,17 @@
+import * as els from "../htmlRenderer.js"
+
+export function introduction(recipe) {
+  const firstParagraphElement = els.p({innerHTML: recipe.description[0]});
+  const firstParagraphDiv = els.div({className: "first-paragraph", children: [firstParagraphElement]});
+  const latterDescriptionParagraphs = recipe.description.slice(1);
+
+  const latterParagraphElements = latterDescriptionParagraphs.map(paragraph => {
+    return els.p({innerHTML: paragraph});
+  });
+
+  const children = [firstParagraphDiv].concat(latterParagraphElements);
+
+  render(recipeParentElement, els.div({className: "introduction-paragraphs", children: children}));
```
This is also a good candidate for a component, if it returns without `render`-ing, and lets `recipe.js` do that.

thomcorley replied:
Yes, that would clean up the introduction module nicely, I will make the change!

ShaneDrury commented on js/recipe/introduction.js:
```diff
@@ -0,0 +1,17 @@
+    return els.p({innerHTML: paragraph});
+  });
+
+  const children = [firstParagraphDiv].concat(latterParagraphElements);
```
This can also be written as:

```js
const children = [firstParagraphDiv, ...latterParagraphElements];
```

thomcorley replied:
Ah yes, that is neater. Thank you!

ShaneDrury commented on js/recipe/introduction.js:
```diff
@@ -0,0 +1,17 @@
+  const firstParagraphDiv = els.div({className: "first-paragraph", children: [firstParagraphElement]});
+  const latterDescriptionParagraphs = recipe.description.slice(1);
+
+  const latterParagraphElements = latterDescriptionParagraphs.map(paragraph => {
+    return els.p({innerHTML: paragraph});
+  });
```
This can be written as:
```js
const latterParagraphElements = latterDescriptionParagraphs.map(paragraph => els.p({innerHTML: paragraph}));
```

where we have an implicit return value from the map function.

thomcorley replied:
Yep, I figured out that adding the additional curly braces meant that the return was required. I'm still learning the basics of es6 syntax!

thomcorley committed:
Use declarative approach in all modules

All modules are now using a more-declarative approach in which all the
information to create the whole section (eg: ingredients list, method
steps etc) is being returned as an object by the module, and is then
being rendered in recipe.js, with the parent element being passed in.

There is one exception; metadata.js, which still renders more than once
within the module. This is due to some complexity with sibling elements
and will be addressed in a future PR.

thomcorley committed:
Merge branch 'master' into declarative-refactoring
